src/datadog/grok/matchers/date.rs

```diff
@@ -241,76 +243,122 @@ pub fn time_format_to_regex(format: &str, with_captures: bool) -> Result<RegexRe
 }
 
 pub fn apply_date_filter(value: &Value, filter: &DateFilter) -> Result<Value, GrokRuntimeError> {
+    // we might have to correct the strp format
+    let mut strp_format = filter.strp_format.clone();
+
+    let year_is_missing =
+        !filter.original_format.contains('y') && !filter.original_format.contains('Y');
+    let month_is_missing = !filter.original_format.contains('M');
+    let day_is_missing = !strp_format.contains('d');
+
+    if day_is_missing {
+        strp_format = format!("%-d {}", strp_format);
+    }
+    if month_is_missing {
+        strp_format = format!("%-m {}", strp_format);
+    }
+    if year_is_missing {
+        strp_format = format!("%Y {}", strp_format);
+    }
```

Comment:
I don't see any bugs here, however it is a complex piece of code. We can improve readability/maintainability by extracting this in a separate function e.g. 

```
pub fn adjust_strp_format(
    original_format: &str, 
    filter_format: &str
) -> String {
    let mut adjusted_format = filter_format.to_string();

    let year_is_missing = !original_format.contains('y') && !original_format.contains('Y');
    let month_is_missing = !original_format.contains('M');
    let day_is_missing = !filter_format.contains('d');

    if day_is_missing {
        adjusted_format = format!("%-d {}", adjusted_format);
    }
    if month_is_missing {
        adjusted_format = format!("%-m {}", adjusted_format);
    }
    if year_is_missing {
        adjusted_format = format!("%Y {}", adjusted_format);
    }

    adjusted_format
}
```

And then we can add a few tests e.g.
```
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_adjust_strp_format() {
        assert_eq!(adjust_strp_format("", strp_format), "%-d %-m %Y ");
        // ...
    }
}
```

Reply:
Good point, updated.